### PR TITLE
`gravatar-ui`- Screenshot testing - Roborazzi

### LINKED ISSUE #114: `gravatar-ui`- Screenshot testing
Review and try to implement **a stable** set of screenshot tests.

### PR DESCRIPTION
**Pull request by hamorillo**
Closes #114 

### Description

This PR adds screenshot testing using [Roborazzi](https://github.com/takahirom/roborazzi). At the moment, screenshots have been added to the `gravatar-ui` module to verify that we don't break anything between iterations of the UI components.

If you execute the verification task, you'll find the report in:

`gravatar-ui/build/reports/roborazzi/index.html`

**Note:** We should add a placeholder to the `Avatar` component in order to prevent it from showing nothing. Another option could be adding the skeleton while the image is being loaded. I've tried but encountered some issues with the screenshots, so I'll leave it for another PR.

### Testing Steps

- Review CI and verify tests have been executed
- Execute locally the `record` and `verify` commands:
```
- Record -> ./gradlew recordRoborazziDebug
- Verify -> ./gradlew verifyRoborazziDebug
```
- Play with the test, trying to break something on purpose to verify it works.

### CODE REVIEW

File: .buildkite/pipeline.yml [hunk 1]
```diff
@@ -26,10 +26,10 @@ steps:
       # `build` task would also run tasks, which are executed in a separate job
       ./gradlew assemble
 
-  - label: "Unit tests"
+  - label: "Unit and screenshot tests"
     command: |
-      echo "--- 🧪 Testing"
-      ./gradlew test
+      echo "--- 🧪📸 Testing"
+      ./gradlew verifyRoborazziDebug
```
Comment (hamorillo) [inline .buildkite/pipeline.yml]: All unit tests will be executed by the `verifyRoborazziDebug` task, not only screenshot ones.

File: gravatar-ui/src/test/java/com/gravatar/gravatar/ui/RoborazziTest.kt [hunk 2]
```diff
@@ -0,0 +1,43 @@
...
+import com.gravatar.ui.GravatarTheme
+import org.junit.Rule
+import org.junit.runner.RunWith
+import org.robolectric.annotation.Config
+import org.robolectric.annotation.GraphicsMode
+
+@RunWith(AndroidJUnit4::class)
+@GraphicsMode(GraphicsMode.Mode.NATIVE)
+@Config(qualifiers = RobolectricDeviceQualifiers.Pixel5)
+abstract class RoborazziTest {
```
Comment (hamorillo) [inline gravatar-ui/src/test/java/com/gravatar/gravatar/ui/RoborazziTest.kt:21]: All screenshot test classes should extend from `RoborazziTest`.

File: README.md [hunk 3]
```diff
@@ -43,6 +43,20 @@ Run unit tests on your machine via the following command:
 ./gradlew test
```
Comment (maxme) [inline README.md:43]: Roborazzi tests fail when I run `./gradlew test`:

> com.gravatar.gravatar.ui.components.LargeProfileSummaryTest > profileLoadingLight FAILED
>    java.lang.RuntimeException at RoboMonitoringInstrumentation.java:105

We should change the README or exclude Roborazzi tests from being ran by `./gradlew test`. I think the second option is better.

File: .buildkite/pipeline.yml (hunk 1 above)
Comment (maxme) [reply .buildkite/pipeline.yml]: I tried to break one of the test in `AvatarUrlTest` and then I ran `./gradlew verifyRoborazziDebug`, I got a `BUILD SUCCESSFUL in 6s`, so I think our unit tests are not run by the `verifyRoborazziDebug` task.

Note that `AvatarUrlTest` and other unit tests are under the `gravatar` module, not `gravatar-ui`.

File: README.md (hunk 3 above)
Comment (maxme) [reply README.md:43]: We only have Roborazzi tests under `gravatar-ui`, so we could make a difference and tell users to run `:gravatar:test` for other unit test and `./gradlew :gravatar-ui:verifyRoborazziDebug` for checking screenshot tests.

**Review by maxme — CHANGES_REQUESTED** [review]
We can't change the `.buildkite/pipeline.yml` that way as the unit tests from the `gravatar` module won't be run, we can eventually add a new task to run `./gradlew :gravatar:test` only

File: .buildkite/pipeline.yml (hunk 1 above)
Comment (maxme) [reply .buildkite/pipeline.yml]: One way to update it is to change this to `./gradlew :gravatar:test ` and add a new task to run `./gradlew :gravatar-ui:verifyRoborazziDebug`

File: gradle.properties [hunk 4]
```diff
@@ -24,3 +24,6 @@ android.nonTransitiveRClass=true
 
 org.gradle.caching=true
 org.gradle.configuration-cache=true
+
+# Roborazzi - https://github.com/takahirom/roborazzi?tab=readme-ov-file#q-how-can-i-execute-screenshot-tests-using-android-studios-run-button
+roborazzi.test.verify=true
```
Comment (hamorillo) [inline gradle.properties:29]: This allows us to run the screenshot tests from the ID by tapping the `run` button.

**Comment by hamorillo** [comment]
Great catch @maxme! Thank you for your in-depth review.

I've excluded all the screenshot tests from the default tests task. Now, when executing `./gradlew test` only non-screenshot tests will be executed. To be more specific, tests that are not annotated with `@ScreenshotTest`. 

If we want to execute screenshot tests, we need to add to the gradlew task the `-Pscreenshot` property. 

Ideally, I would like to remove the necessity of the `-Pscreenshot` property. I'll iterate over it, but I don't think it's a blocker right now.

Wdty?

**Comment by maxme** [comment]
> Ideally, I would like to remove the necessity of the `-Pscreenshot` property. I'll iterate over it, but I don't think it's a blocker right now.

That's fine, if we use them mostly for catching regressions, it will be fine if it's correctly set up in CI.

**Review by maxme — APPROVED** [review]
- `./gradlew :gravatar-ui:verifyRoborazziDebug -Pscreenshot` work as expected and run the screenshot tests.
- `./gradle test` work as expected, run the unit test and doesn't run the screenshot tests.


LGTM - :shipit:

**Pull request merged**